### DISCUSSION TITLE
Example of throwing error in /server/api using h3?

### DISCUSSION THREAD

**Discussion by justindasilva** [nuxt/nuxt#16908] [Questions]
Are there any full examples of `h3` being used within the `/server/api` directory?

I'm not quite sure how to send an error.

It seems I shouldn't be throwing an error.
And `res.sendError` didn't work.

I want to be able to catch errors on the frontend using `$nuxt/http`

I created a plugin with the following:
```
export default ({ $http, redirect }) => {

  $http.onError(error => {

    if (error.statusCode === 500) {
      $nuxt.$bvToast.toast(error.response.data.error, {
        title: error.message,
        autoHideDelay: 5000,
        solid: true,
        variant: 'danger',
      })
    }

})

}
```

**Comment by danielroe** [comment]
You can just throw an error, return an error, or use the `createError` utility from `h3`.

**Reply by justindasilva** [reply]
If I simply throw an error in the endpoint, I will get this response when hitting it:
```
<!DOCTYPE html><html><head><title>500 - Internal Server Error | Nuxt</title>
...
    <h1>Field missing</h1>
```

Is there a way to hook into the functionality behind `/server/api` and return a json error response when I throw an error? Throwing errors in my code is my preferred way to return an error response.

For example, when I was using express in Nuxt 2, I had done this:
```
const app = express()

const errorHandler = (err, req, res, next) => {
  if (res.headersSent) {
    return next(err)
  } else {
    console.log(err)
    res.status(err.status || 500);
    res.send({
      error: err.message || 'An error has occured',
    });
  }
}

app.use(errorHandler)
```

In the meantime, I'll take a look at `createError`

**Reply by danielroe** [reply]
If you visit it in the browser, you'll get an HTML error page. But if you get it via a fetch request it will be JSON. (Assuming you have header `Accept: application/json`.)

**Reply by justindasilva** [reply]
This definitely got me a step further in the right direction. I'm now getting back JSON, so thanks for that!

However, if I `throw new Error('test')`, I get the following JSON response:
```
{"statusCode":500,"statusMessage":"Internal Server Error","description":"\n    <h1>test</h1>\n    <pre><span class=\"stack\">at all$2 (file://./.nuxt/nitro/index.mjs:506:9)</span>\n<span class=\"stack internal\">at file://./node_modules/h3/dist/index.mjs:111:24</span>\n<span class=\"stack internal\">at new Promise (<anonymous>)</span>\n<span class=\"stack internal\">at callHandle (file://./node_modules/h3/dist/index.mjs:108:10)</span>\n<span class=\"stack internal\">at file://./node_modules/h3/dist/index.mjs:104:12</span>\n<span class=\"stack internal\">at file://./node_modules/h3/dist/index.mjs:132:34</span>\n<span class=\"stack internal\">at processTicksAndRejections (node:internal/process/task_queues:96:5)</span>\n<span class=\"stack internal\">at async handle (file://./node_modules/h3/dist/index.mjs:572:19)</span></pre>\n    "}
```

How can I get `test` in the `statusMessage`?
Either that, or can I add a new parameter `errorMessage` that has the value I can use on the frontend?

Even if I do `throw createError({ statusCode: 500, statusMessage: 'test' })`, this still doesn't override the `statusMessage` properly.
I can't help but wonder if this is a bug?

**Reply by RamiAwar** [reply]
Worked perfectly for me:
```javascript
throw createError({
            statusCode: 404,
            name: 'NotFoundError',
            message: 'Todo not found',
            statusMessage: "Not Found",
        })
```

Response:
```json
{
    "url": "/api/todo/aca2e3ac-0b0b-4e22-9a20-8b26b48f53bc",
    "statusCode": 404,
    "statusMessage": "Not Found",
    "message": "Todo not found",
    "stack": ""
}
```

**Comment by cdwmhcc** [comment]
> I want to be able to catch errors on the frontend using `$nuxt/http`

Did you solve it?

**Comment by mklueh** [comment]
How would you throw a 401 if a user is not authenticated?

Examples in the docs would really be amazing


Edit:


Found it, you need to do

```
import {sendError, createError} from "h3"

export default defineEventHandler((event) => {
  if (event.req.originalUrl === '/api/posts')
    sendError(event, createError({statusCode: 401, statusMessage: 'Unauthenticated'}))
})
```


**Why are the middleware handlers applied to page loads even if the ssr mode is set to false?**

```
export default defineNuxtConfig({
  ssr: false,


})
```

**If I don't do explicitly   `if (event.req.originalUrl === '/api/posts')` I cannot access any page. Is it possible to apply the auth handler just to the REST API?**
